### ISSUE TITLE
Python autograder in PL will not continue after encountering an ERROR

### ISSUE THREAD

**Issue by celinaberg** [PrairieLearn/PrairieLearn#5491]
I am noticing the Python autograder in PL will not continue if an exception is thrown by the student code within a test.
I know the Python unittest does not behave this way, it will run through all tests even if there is an ERROR or FAIL in one of the earlier tests.
With the existing PL implementation, it won’t allocate points for tests that would have run and passed after the exception is thrown.
To add to this, the reported Score become inconsistent:
`Score: 4/5 (66.67%)`
notice: the fraction shown is the passed tests/ the number of tests that ran before the exception was thrown
but the % grade is the passed tests/ the number of tests in the unit tester

**Comment by reteps** [comment]
From Claude, discussion about a proposed fix:

- Can’t just change done_grading to skip_grading — Python unittest blocks test execution before PrairieLearn’s flags are even checked.

- Can’t route through existing addError(UserCodeFailedError) — that path was designed for _ErrorHolder (where startTest was never called) and appends a phantom result entry. When startTest HAS been called, you get double entries and a KeyError crash.

- Must handle it directly in run() — call startTest, set result.results[-1]["points"] = 0, write feedback, return. No addError call.

- Must set cls.ref = cls.st = cls.plt = None in the except block — otherwise tearDownClass crashes with AttributeError (especially for PLTestCaseWithPlot).

- SyntaxError/NameError needs a gate — only process it once (first test), using done_grading as the gate, and don’t register any test entries (matches current ungradable behavior).

- Non-SyntaxError errors: show the full traceback on the first test, show “skipped” on subsequent tests. All tests appear in the grading panel with 0 points. Score is 0 / total — consistent.

**Comment by nwalters512** [comment]
@eliotwrobson does your revamped Python grader account for this?

**Comment by eliotwrobson** [comment]
Yes, following the default behavior by pytest, the grader runs all test cases even if there's an exception or other unexpected failure during a test case.
